File: python/exercises/practice/unit_test_9db9374e_4935_44d9_b1c2_f8e7f7eaf524/trade_network/.meta/trade_network.py
import heapq
from collections import defaultdict

def min_cost_flow(num_nodes, edges, source, destination, amount):
    """
    Calculate the minimum cost to route the required amount of trade from source to destination.
    
    Args:
        num_nodes: Number of trading hubs (nodes).
        edges: List of tuples (source_node, destination_node, capacity, cost).
        source: Index of the source trading hub.
        destination: Index of the destination trading hub.
        amount: Total amount of trade to be routed.
    
    Returns:
        Minimum total cost to route the required amount of trade, or -1.0 if impossible.
    """
    # Special case: source and destination are the same
    if source == destination:
        return 0.0
    
    # Build the graph as an adjacency list
    graph = defaultdict(list)
    for u, v, cap, cost in edges:
        if cap > 0:  # Only consider edges with positive capacity
            graph[u].append((v, cap, cost))
    
    total_flow = 0.0
    total_cost = 0.0
    
    # Continue finding shortest paths and augmenting flow until we meet the demand
    while total_flow < amount:
        # Use Dijkstra's algorithm to find the shortest path from source to destination
        dist = [float('inf')] * num_nodes
        dist[source] = 0
        prev = [-1] * num_nodes
        prev_edge = [-1] * num_nodes
        pq = [(0, source)]
        
        while pq and dist[destination] == float('inf'):
            cost_so_far, node = heapq.heappop(pq)
            
            if cost_so_far > dist[node]:
                continue
            
            for i, (neighbor, capacity, cost) in enumerate(graph[node]):
                if capacity > 0 and dist[node] + cost < dist[neighbor]:
                    dist[neighbor] = dist[node] + cost
                    prev[neighbor] = node
                    prev_edge[neighbor] = i
                    heapq.heappush(pq, (dist[neighbor], neighbor))
        
        # If no path found, we can't route more flow
        if dist[destination] == float('inf'):
            break
        
        # Find maximum flow that can be pushed through the found path
        path_flow = amount - total_flow
        node = destination
        while node != source:
            parent = prev[node]
            edge_idx = prev_edge[node]
            edge_capacity = graph[parent][edge_idx][1]
            path_flow = min(path_flow, edge_capacity)
            node = parent
        
        # Update capacities along the path
        node = destination
        while node != source:
            parent = prev[node]
            edge_idx = prev_edge[node]
            # Reduce capacity of forward edge
            v, cap, edge_cost = graph[parent][edge_idx]
            graph[parent][edge_idx] = (v, cap - path_flow, edge_cost)
            
            # Update total cost
            total_cost += path_flow * edge_cost
            
            node = parent
        
        total_flow += path_flow
    
    # Check if we managed to route the entire amount
    if total_flow < amount:
        return -1.0
    
    return total_cost

def min_cost_flow_network_simplex(num_nodes, edges, source, destination, amount):
    """
    Alternative implementation using Network Simplex algorithm,
    which is more efficient for min-cost flow problems.
    
    Note: This implementation is more complex and is provided as a more
    efficient alternative to the augmenting path approach.
    
    Args:
        num_nodes: Number of trading hubs (nodes).
        edges: List of tuples (source_node, destination_node, capacity, cost).
        source: Index of the source trading hub.
        destination: Index of the destination trading hub.
        amount: Total amount of trade to be routed.
    
    Returns:
        Minimum total cost to route the required amount of trade, or -1.0 if impossible.
    """
    # Special case: source and destination are the same
    if source == destination:
        return 0.0
    
    # Initialize network with a super source and super sink
    # This allows us to handle the problem as a circulation problem
    n = num_nodes + 2
    super_source = num_nodes
    super_sink = num_nodes + 1
    
    # Build the graph with costs and capacities
    graph = [[] for _ in range(n)]
    cost = {}
    capacity = {}
    flow = {}
    
    # Add original edges
    for u, v, cap, c in edges:
        if cap > 0:  # Only consider edges with positive capacity
            graph[u].append(v)
            if (u, v) not in capacity:
                capacity[(u, v)] = 0
                cost[(u, v)] = 0
                flow[(u, v)] = 0
            capacity[(u, v)] += cap
            # Use weighted average for cost if there are parallel edges
            cost[(u, v)] = (cost[(u, v)] * flow[(u, v)] + c * cap) / (flow[(u, v)] + cap)
    
    # Add edges from super_source to source and from destination to super_sink
    graph[super_source].append(source)
    capacity[(super_source, source)] = amount
    cost[(super_source, source)] = 0
    flow[(super_source, source)] = 0
    
    graph[destination].append(super_sink)
    capacity[(destination, super_sink)] = amount
    cost[(destination, super_sink)] = 0
    flow[(destination, super_sink)] = 0
    
    # Add return edge from super_sink to super_source to make it a circulation
    graph[super_sink].append(super_source)
    capacity[(super_sink, super_source)] = amount
    cost[(super_sink, super_source)] = 0
    flow[(super_sink, super_source)] = 0
    
    # Initialize potential for each node
    potential = [0] * n
    
    # Initialize spanning tree
    # For simplicity, we'll use an augmenting path approach here
    # A full implementation of Network Simplex would maintain an explicit tree structure
    
    # Find an initial feasible flow (e.g., using augmenting paths)
    # For simplicity, we'll use Edmonds-Karp algorithm
    def bfs(s, t):
        visited = [False] * n
        queue = [s]
        visited[s] = True
        prev = {}
        
        while queue and not visited[t]:
            node = queue.pop(0)
            for neighbor in graph[node]:
                if not visited[neighbor] and capacity.get((node, neighbor), 0) > flow.get((node, neighbor), 0):
                    queue.append(neighbor)
                    visited[neighbor] = True
                    prev[neighbor] = (node, True)  # True indicates a forward edge
            
            # Also consider residual edges (backward edges)
            for u in range(n):
                if (u, node) in flow and flow[(u, node)] > 0 and not visited[u]:
                    queue.append(u)
                    visited[u] = True
                    prev[u] = (node, False)  # False indicates a backward edge
        
        if visited[t]:
            return prev
        return None
    
    # Find max flow using augmenting paths
    while True:
        prev = bfs(super_source, super_sink)
        if not prev:
            break
        
        # Find the bottleneck capacity
        path_flow = float('inf')
        node = super_sink
        while node != super_source:
            parent, is_forward = prev[node]
            if is_forward:
                path_flow = min(path_flow, capacity.get((parent, node), 0) - flow.get((parent, node), 0))
            else:
                path_flow = min(path_flow, flow.get((node, parent), 0))
            node = parent
        
        # Update flow along the path
        node = super_sink
        while node != super_source:
            parent, is_forward = prev[node]
            if is_forward:
                flow[(parent, node)] = flow.get((parent, node), 0) + path_flow
            else:
                flow[(node, parent)] = flow.get((node, parent), 0) - path_flow
            node = parent
    
    # Check if we managed to route the entire amount
    if flow.get((super_source, source), 0) < amount:
        return -1.0
    
    # Calculate the total cost
    total_cost = 0.0
    for (u, v), f in flow.items():
        if u != super_source and u != super_sink and v != super_source and v != super_sink:
            total_cost += f * cost.get((u, v), 0)
    
    return total_cost

# By default, use the simpler augmenting path approach
# The more efficient Network Simplex approach is provided as an alternative

File: go/exercises/practice/unit_test_38183764_6b58_434e_8f9e_39376b734694/trade_network/.meta/trade_network.go
package trade_network

import (
	"container/heap"
	"math"
)

type edge struct {
	to, rev, cap, cost int
}

type mcmf struct {
	n   int
	g   [][]edge
	pot []int
	dist []int
	prevV []int
	prevE []int
}

func newMCMF(n int) *mcmf {
	return &mcmf{
		n:    n,
		g:    make([][]edge, n),
		pot:  make([]int, n),
		dist: make([]int, n),
		prevV: make([]int, n),
		prevE: make([]int, n),
	}
}

func (m *mcmf) addEdge(from, to, cap, cost int) {
	m.g[from] = append(m.g[from], edge{to: to, rev: len(m.g[to]), cap: cap, cost: cost})
	m.g[to] = append(m.g[to], edge{to: from, rev: len(m.g[from]) - 1, cap: 0, cost: -cost})
}

type item struct {
	node, dist int
	index int
}

type priorityQueue []*item

func (pq priorityQueue) Len() int { return len(pq) }
func (pq priorityQueue) Less(i, j int) bool { return pq[i].dist < pq[j].dist }
func (pq priorityQueue) Swap(i, j int) { 
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *priorityQueue) Push(x interface{}) {
	n := len(*pq)
	it := x.(*item)
	it.index = n
	*pq = append(*pq, it)
}

func (pq *priorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	it := old[n-1]
	old[n-1] = nil
	it.index = -1
	*pq = old[0 : n-1]
	return it
}

func (m *mcmf) minCostFlow(s, t, f int) (int, int) {
	res := 0
	inf := math.MaxInt32
	// initialize potential to 0 for all nodes
	for i := 0; i < m.n; i++ {
		m.pot[i] = 0
	}
	flow := 0
	for f > 0 {
		// Dijkstra to find shortest path from s to t in residual graph
		for i := 0; i < m.n; i++ {
			m.dist[i] = inf
		}
		m.dist[s] = 0
		pq := &priorityQueue{}
		heap.Init(pq)
		heap.Push(pq, &item{node: s, dist: 0})
		for pq.Len() > 0 {
			it := heap.Pop(pq).(*item)
			v := it.node
			if m.dist[v] < it.dist {
				continue
			}
			for i, e := range m.g[v] {
				if e.cap > 0 && m.dist[e.to] > m.dist[v] + e.cost + m.pot[v] - m.pot[e.to] {
					m.dist[e.to] = m.dist[v] + e.cost + m.pot[v] - m.pot[e.to]
					m.prevV[e.to] = v
					m.prevE[e.to] = i
					heap.Push(pq, &item{node: e.to, dist: m.dist[e.to]})
				}
			}
		}
		if m.dist[t] == inf {
			break
		}
		// update potentials for all nodes
		for v := 0; v < m.n; v++ {
			if m.dist[v] < inf {
				m.pot[v] += m.dist[v]
			}
		}
		// add as much flow as possible along the found path
		addFlow := f
		for v := t; v != s; v = m.prevV[v] {
			if addFlow > m.g[m.prevV[v]][m.prevE[v]].cap {
				addFlow = m.g[m.prevV[v]][m.prevE[v]].cap
			}
		}
		f -= addFlow
		flow += addFlow
		res += addFlow * m.pot[t]
		for v := t; v != s; v = m.prevV[v] {
			e := &m.g[m.prevV[v]][m.prevE[v]]
			e.cap -= addFlow
			m.g[v][e.rev].cap += addFlow
		}
	}
	return flow, res
}

// MinCostTrade computes the minimal total cost of trading goods across the network.
// It processes each good individually and sums the cost.
func MinCostTrade(numPlanets int, numGoods int, demandSupply [][]int, wormholes [][]int, penalty int) int {
	totalCost := 0
	// For each good, build network and compute min cost flow.
	for g := 0; g < numGoods; g++ {
		n := numPlanets + 2
		S := numPlanets
		T := numPlanets + 1
		mcmfNet := newMCMF(n)
		totalDemand := 0
		// For each planet, add supply edge if negative supply, and demand edge if positive.
		for i := 0; i < numPlanets; i++ {
			val := demandSupply[i][g]
			if val < 0 {
				// Supply available: add edge from S -> i with capacity = -val.
				mcmfNet.addEdge(S, i, -val, 0)
			} else if val > 0 {
				// Demand: add edge from i -> T with capacity = val.
				mcmfNet.addEdge(i, T, val, 0)
				totalDemand += val
			}
		}
		// Add penalty edge from S -> T to allow unmet demand at cost penalty per unit.
		mcmfNet.addEdge(S, T, totalDemand, penalty)
		// Add wormhole edges for this good.
		// Each wormhole format: [source, destination, cost, cap_good0, cap_good1, ..., cap_good_(numGoods-1)]
		for _, wh := range wormholes {
			if len(wh) < 3+numGoods {
				continue
			}
			u := wh[0]
			v := wh[1]
			cost := wh[2]
			capacity := wh[3+g]
			if capacity > 0 {
				mcmfNet.addEdge(u, v, capacity, cost)
			}
		}
		// The required flow is totalDemand.
		_, cost := mcmfNet.minCostFlow(S, T, totalDemand)
		totalCost += cost
	}
	return totalCost
}